File: Code/js/Lesson_16_code/src/staticSite.ts
// The TypeScript twin of Code/Lesson_16_code/static_site.py — the static
// origin, http://127.0.0.1:8051. Serves index.html, style.css, and
// script.js from src/static/.
//
//   CSP_MODE=on          add a Content-Security-Policy header to the HTML
//   node src/staticSite.ts    listens on 127.0.0.1:8051, plain HTTP

import http from "node:http";
import { readFile } from "node:fs/promises";
import path from "node:path";
import { fileURLToPath } from "node:url";

const LISTEN_HOST = "127.0.0.1";
const LISTEN_PORT = 8051;
const CSP_MODE = process.env.CSP_MODE === "on";
const ROOT = path.join(path.dirname(fileURLToPath(import.meta.url)), "static");

const CONTENT_TYPES: Record<string, string> = {
  ".html": "text/html; charset=utf-8",
  ".js": "text/javascript; charset=utf-8",
  ".css": "text/css; charset=utf-8",
};

const server = http.createServer(async (req, res) => {
  console.log(`staticSite: ${req.socket.remoteAddress} ${req.method} ${req.url}`);
  const rel = req.url === "/" ? "index.html" : (req.url ?? "").replace(/^\//, "");
  const filePath = path.join(ROOT, rel);
  try {
    const body = await readFile(filePath);
    const ext = path.extname(filePath);
    const headers: http.OutgoingHttpHeaders = {
      "Content-Type": CONTENT_TYPES[ext] ?? "application/octet-stream",
      "Content-Length": body.length,
      Connection: "close",
    };
    if (CSP_MODE && ext === ".html") {
      headers["Content-Security-Policy"] = "default-src 'self'";
    }
    res.writeHead(200, headers);
    res.end(body);
  } catch {
    res.writeHead(404, { Connection: "close" });
    res.end();
  }
});

server.listen(LISTEN_PORT, LISTEN_HOST, () => {
  console.log(`staticSite listening on http://${LISTEN_HOST}:${LISTEN_PORT} (CSP_MODE=${CSP_MODE})`);
});

File: Code/Lesson_16_code/static_site.py
"""The static origin: http://127.0.0.1:8051.

A second origin, on a different port from api.py, serving one HTML page and
one script file. Nothing here talks to a database; the point is only that
this origin's scheme, host, and port differ from the API's, which is enough
to make the browser treat every request between them as cross-origin.

    CSP_MODE=on    add a Content-Security-Policy header to the HTML response
    python3 static_site.py    listens on 127.0.0.1:8051, plain HTTP
"""

from __future__ import annotations

import os
from http.server import BaseHTTPRequestHandler, ThreadingHTTPServer
from pathlib import Path

LISTEN_HOST = "127.0.0.1"
LISTEN_PORT = 8051
CSP_MODE = os.environ.get("CSP_MODE") == "on"
ROOT = Path(__file__).parent / "static"

CONTENT_TYPES = {
    ".html": "text/html; charset=utf-8",
    ".js": "text/javascript; charset=utf-8",
    ".css": "text/css; charset=utf-8",
}


class Handler(BaseHTTPRequestHandler):
    protocol_version = "HTTP/1.1"

    def log_message(self, format: str, *args: object) -> None:  # noqa: A002
        print(f"static_site: {self.address_string()} {format % args}", flush=True)

    def do_GET(self) -> None:
        rel = "index.html" if self.path == "/" else self.path.lstrip("/")
        path = ROOT / rel
        if not path.is_file():
            self.send_response(404)
            self.send_header("Connection", "close")
            self.end_headers()
            self.close_connection = True
            return
        body = path.read_bytes()
        self.send_response(200)
        self.send_header("Content-Type", CONTENT_TYPES.get(path.suffix, "application/octet-stream"))
        self.send_header("Content-Length", str(len(body)))
        if CSP_MODE and path.suffix == ".html":
            self.send_header("Content-Security-Policy", "default-src 'self'")
        self.send_header("Connection", "close")
        self.end_headers()
        self.wfile.write(body)
        self.close_connection = True


if __name__ == "__main__":
    server = ThreadingHTTPServer((LISTEN_HOST, LISTEN_PORT), Handler)
    print(f"static_site listening on http://{LISTEN_HOST}:{LISTEN_PORT} (CSP_MODE={CSP_MODE})", flush=True)
    server.serve_forever()
